### ISSUE TITLE
Check for None values when constructing R command-line invocations

### ISSUE THREAD

**Issue by mikej888** [riboviz/RiboViz#72]
Reported by @FlicAnderson in relation to #66.

If the YAML configuration file contains:

```yaml
features_file: data/yeast_features.tsv # file of features to correlate with ORFs
do_pos_sp_nt_freq: TRUE # Do calculate position-specific nucleotide freqeuency
t_rna: data/yeast_tRNAs.tsv # tRNA estimates
codon_pos: data/yeast_codon_pos_i200.RData # Codon positions in each gene
```
Running:
```console
$ python -m riboviz.tools.prep_riboviz --dry-run rscripts/ vignette/vignette_config.yaml
```
succeeds and prep_riboviz.sh contains:
```
Rscript --vanilla rscripts/generate_stats_figs.R --Ncores=1 --MinReadLen=10 
--MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette 
--hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT 
--orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True 
--orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 
--dir_out=vignette/output --t_rna=data/yeast_tRNAs.tsv 
--codon_pos=data/yeast_codon_pos_i200.RData 
--features_file=data/yeast_features.tsv --do_pos_sp_nt_freq=True
```
If the YAML configuration file contains:
```yaml
features_file: NULL # file of features to correlate with ORFs
do_pos_sp_nt_freq: FALSE # Do calculate position-specific nucleotide freqeuency
t_rna: NULL # tRNA estimates
codon_pos: NULL # Codon positions in each gene
```
Running:
```console
$ python -m riboviz.tools.prep_riboviz --dry-run rscripts/ vignette/vignette_config.yaml
```
fails with:
```
Traceback (most recent call last):
  File "/home/ubuntu/RiboViz/riboviz/tools/prep_riboviz.py", line 534, in prep_riboviz
    dry_run)
  File "/home/ubuntu/RiboViz/riboviz/tools/prep_riboviz.py", line 364, in process_sample
    "--t_rna=" + config["t_rna"],
TypeError: can only concatenate str (not "NoneType") to str
```
The problem arises as the configuration value for t_rna is value None in Python.

**Comment by mikej888** [comment]
Created branch check-none-72. 

Commit 2e4e2a2479016c9665a1915689b4608d1b60fa2b checks if any configuration value is `None` in Python (`NULL` in YAML) and, if so, does not add it.

Now if the YAML configuration file contains:

```yaml
features_file: data/yeast_features.tsv # file of features to correlate with ORFs
do_pos_sp_nt_freq: TRUE # Do calculate position-specific nucleotide freqeuency
t_rna: data/yeast_tRNAs.tsv # tRNA estimates
codon_pos: data/yeast_codon_pos_i200.RData # Codon positions in each gene
```
Running:
```console
$ python -m riboviz.tools.prep_riboviz --dry-run rscripts/ vignette/vignette_config.yaml
```
succeeds and prep_riboviz.sh contains:
```
Rscript --vanilla rscripts/generate_stats_figs.R --Ncores=1 --MinReadLen=10 --MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette --hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT --orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --dir_out=vignette/output --do_pos_sp_nt_freq=True --t_rna=data/yeast_tRNAs.tsv --codon_pos=data/yeast_codon_pos_i200.RData --features_file=data/yeast_features.tsv --orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3
```
If the YAML configuration file contains:
```yaml
features_file: NULL # file of features to correlate with ORFs
do_pos_sp_nt_freq: FALSE # Do calculate position-specific nucleotide freqeuency
t_rna: NULL # tRNA estimates
codon_pos: NULL # Codon positions in each gene
```
Running:
```console
$ python -m riboviz.tools.prep_riboviz --dry-run rscripts/ vignette/vignette_config.yaml
```
also succeeds and prep_riboviz.sh contains:
```
Rscript --vanilla rscripts/generate_stats_figs.R --Ncores=1 --MinReadLen=10 --MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette --hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT --orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --dir_out=vignette/output --do_pos_sp_nt_freq=False --orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3
```

**Comment by mikej888** [comment]
Running the following in "develop"
```R
Rscript --vanilla rscripts/generate_stats_figs.R --Ncores=1 --MinReadLen=10 
--MaxReadLen=50 --Buffer=250 --PrimaryID=Name --dataset=vignette 
--hdFile=vignette/output/WT3AT.h5 --out_prefix=vignette/output/WT3AT 
--orf_fasta=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True 
--dir_out=vignette/output --do_pos_sp_nt_freq=False 
--orf_gff_file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3
```
shows no configuration values within R defined for the missing parameters:
```
$orf_gff_file
[1] "vignette/input/yeast_YAL_CDS_w_250utrs.gff3"

$do_pos_sp_nt_freq
[1] FALSE
```
@FlicAnderson noted this is due to the definition of the parameters being `NULL` in `generate_stats_figs.R`:

```
  make_option("--t_rna", type="character", default=NULL,
              help="tRNA estimates in .tsv file"),
  make_option("--codon_pos", type="character", default=NULL,
              help="Codon positions in each gene in .Rdata file"),
  make_option("--features_file", type="character", default=NULL,
              help="features file, columns are gene features and rows are genes"),
```

Running it in a branch of handle-missing-features-file-66 (see #66) into which check-none-72 has been merged gives:

```
$t_rna
[1] NA

$codon_pos
[1] NA

$orf_gff_file
[1] "vignette/input/yeast_YAL_CDS_w_250utrs.gff3"

$features_file
[1] NA

$do_pos_sp_nt_freq
[1] FALSE
```
This is because @FlicAnderson updated `generate_stats_figs.R` to define the default value for these parameters as `NA`:

```R
  make_option("--t_rna", type="character", default=NA,
              help="tRNA estimates in .tsv file"),
  make_option("--codon_pos", type="character", default=NA,
              help="Codon positions in each gene in .Rdata file"),
  make_option("--orf_gff_file", type="character", default=NA,
              help="riboviz generated GFF2/GFF3 annotation file"),
  make_option("--features_file", type="character", default=NA,
              help="features file, columns are gene features and rows are genes"),
```

@FlicAnderson: recommend, for #66, you merge check-none-72 into handle-missing-features-file-66 and continue. If you are OK with the changes, please let me know so check-none-72 can be merged into "develop" and this issue closed.

**Comment by mikej888** [comment]
These problems did not arise for `bam_to_h5.R` and its `SecondID` parameter because:

* `SecondID: NULL` is defined in `vignette_config.yaml` so becomes `None` when loaded into `prep_riboviz.py`
* `prep_riboviz.py` checks for it being `None` and converts it to `"NULL"` when constructing the command-line call to `bam_to_h5.R`

```
    if second_id is None:
        second_id = "NULL"
```
* `bam_to_h5.R` reads this value in as a string and then converts it to `NULL` if the string is `"NULL"`:

```R
if (opt$SecondID == "NULL") {
  # unquote NULL option in SecondID
  SecondID <- NULL
```

`bam_to_h5.R` defines this parameter as:

```R
    make_option("--SecondID", type="character", default=NULL,
        help="Secondary gene IDs to access the data (COX1, EFB1, etc.)"),
```

and uses `is.null(SecondID)` checks. Depending on the outcome of #66 then `bam_to_h5.R` can be updated to:

* Set the default value for `SecondID` to be `NA`.
* Remove the check that `SecondID` is a string `"NULL"`.
* Replace use of `is.null` with `is.na` throughout (as @FlicAnderson does in `generate_stats_figs.R` in 
branch handle-missing-features-file-66.

and `prep_riboviz.py` updated to construct the call to `bam_to_h5.R` in the same way as `generate_stats_figs.R`.

**Comment by kavousan** [comment]
@FlicAnderson: assigned to you as per Mike's suggestion at the end of https://github.com/riboviz/RiboViz/issues/72#issuecomment-533572335

**Comment by mikej888** [comment]
Merged as part of pull request #76
